packages/react/ssr/src/SSR.tsx

```diff
@@ -0,0 +1,53 @@
+// This implementation is heavily inspired by react-aria's implementation
+// See: https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/ssr/src/SSRProvider.tsx
+
+import * as React from 'react';
+
+type SSRContextValue = {
+  prefix: number;
+  current: number;
+};
+
+const defaultSSRContext: SSRContextValue = {
+  prefix: Math.round(Math.random() * 10000000000),
+  current: 0,
+};
+
+const SSRContext = React.createContext<SSRContextValue>(defaultSSRContext);
+
+const SSRProvider: React.FC = (props) => {
+  const currentContext = React.useContext(SSRContext);
+  const isRootSSRProvider = currentContext === defaultSSRContext;
+  const context: SSRContextValue = React.useMemo(
+    () => ({
+      prefix: isRootSSRProvider ? 0 : ++currentContext.prefix,
+      current: 0,
+    }),
+    [isRootSSRProvider, currentContext]
+  );
+
+  return <SSRContext.Provider value={context} {...props} />;
+};
+
+function useSSRSafeId(deterministicId?: string): string {
```

Comment:
Is the plan to use this inside our `useId` hook? As in, we wouldn't really interface with this directly?

Reply:
Yeah exactly, there would be a friendlier api.